eval/topdown.go

```diff
@@ -65,89 +64,125 @@ func (ctx *TopDownContext) Bind(variable opalog.Var, value opalog.Value) *TopDow
 		return nil
 	}
 	cpy := *ctx
-	cpy.Bindings = make(Bindings)
-	for k, v := range ctx.Bindings {
-		cpy.Bindings[k] = plugValue(v, Bindings{variable: value})
-	}
-	cpy.Bindings[variable] = value
+	cpy.Bindings = newHashMap()
+	tmp := newHashMap()
+	tmp.Put(variable, value)
+	ctx.Bindings.Iter(func(k opalog.Value, v opalog.Value) bool {
```

Comment:
The logic here is somewhat specific to bindings. I.e., we create a temporary hash map holding the new variable binding. For each binding in the existing bindings, we plug the new binding in while performing the copy. If hashMap is only used to store bindings, implementing union this way would make sense. I would prefer to hold off on this for now.